exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,235 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
 
 	return ctrl.Result{}, nil
 }
+
+// reconcileMachines reconciles Machines associated with a MachinePool.
+//
+// Note: In the case of MachinePools the machines are created in order to surface in CAPI what exists in the
+// infrastructure while instead on MachineDeployments, machines are created in CAPI first and then the
+// infrastructure is created accordingly.
+// Note: When supported by the cloud provider implementation of the MachinePool, machines will provide a means to interact
+// with the corresponding infrastructure (e.g. delete a specific machine in case MachineHealthCheck detects it is unhealthy).
+func (r *MachinePoolReconciler) reconcileMachines(ctx context.Context, mp *expv1.MachinePool, infraMachinePool *unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	var noKind bool
+	var noSelector bool
+
+	var infraMachineKind string
+	if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineKind, "status", "infrastructureMachineKind"); err != nil {
+		if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+			noKind = true
+		} else {
+			return errors.Wrapf(err, "failed to retrieve infraMachineKind from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
```

Comment:
```suggestion
			return errors.Wrapf(err, "failed to retrieve infraMachineKind from infrastructure provider for MachinePool %s", klog.KObj(mp))
```
nit: same below (context: we started using this consistently in core CAPI)